### PR TITLE
Add ap capability flow to controller, agent, and CI

### PR DESCRIPTION
**Pull request by Galvien**
This pull request adds support for passing Controller Certification AP Capability Query test as defined in section 5.6.1 in the EasyMesh Test Plan v1.0 [1].

First, a shell test flow as described in [1][5.6.1] for the AP Capability Query sent by the controller to the agent is added to `test_flows.sh`.

Then, the controller is updated to support sending the query message,
Using the `WFA_CA` interface to trigger the sending of a CMDU from the controller to the agent.

Lastly, the agent backhaul manager was updated to subscribe and handle the received message, thus passing the test.

Implements #164 👌 
(A proper implementation of #161 )

Note that the current handling of the message by the agent is stub, and will be handled later during agent certification.

### CODE REVIEW

File: controller/src/beerocks/master/wfa_ca.cpp [hunk 1]
```diff
@@ -660,6 +660,16 @@ bool wfa_ca::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx, ieee1905_1::eMessag
         }
         break;
     }
+    case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE: {
+
+        cmdu_header = cmdu_tx.create(g_mid, ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE);
+        if (!cmdu_header) {
+            //TODO: change this to the generic error message format
```
Comment (tomereli) [inline controller/src/beerocks/master/wfa_ca.cpp]: delete the todo

Comment (tomereli) [inline controller/src/beerocks/master/wfa_ca.cpp]: So delete the TODO...

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -1581,6 +1585,11 @@ bool main_thread::handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx
     //       agent certification
     return true;
 }
+bool main_thread::handle_1905_ap_capability_query(ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    LOG(DEBUG) << "Received AP Capability Query message";
```
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Add TODO comment here that explains what needs to be done - send the query to all radio agents. Each agent will then respond with its capabilities, and eventually will be sent to the controller.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h [hunk 3]
```diff
@@ -69,6 +69,8 @@ class main_thread : public btl::transport_socket_thread {
     bool handle_1905_autoconfiguration_response(ieee1905_1::CmduMessageRx &cmdu_rx,
                                                 const std::string &src_mac);
     bool handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_1905_ap_capability_query(ieee1905_1::CmduMessageRx &cmdu_rx);
```
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: not needed for this stage, remove

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 4]
```diff
@@ -476,6 +476,8 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
         return handle_channel_preference_query(sd, cmdu_rx);
     case ieee1905_1::eMessageType::CHANNEL_SELECTION_REQUEST_MESSAGE:
         return handle_channel_selection_request(sd, cmdu_rx);
+    case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE:
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: same here

File: agent/src/beerocks/slave/son_slave_thread.h [hunk 5]
```diff
@@ -283,6 +283,7 @@ class slave_thread : public beerocks::socket_thread {
     bool autoconfig_wsc_add_m1();
     bool handle_channel_preference_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
     bool handle_channel_selection_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_1905_ap_capability_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.h]: and here

**Review by tomereli — CHANGES_REQUESTED** [review]

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h (hunk 3 above)
Comment (Galvien) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: I might be missing out something but I'm pretty sure this is just the header for the `main_thread` function that we're using to log.

File: tools/docker/tests/test_flows.sh [hunk 6]
```diff
@@ -91,14 +91,12 @@ test_channel_selection() {
...
 
-test_ap_capability_info_reporting() {
-    #TODO: Implement
-    return 1
-}
-
-test_client_capability_info_reporting() {
-    #TODO: Implement
-    return 1
+test_ap_capability_query() {
```
Comment (arnout) [inline tools/docker/tests/test_flows.sh:105]: If you change the name, it should also be changed in the help text and in the ALL variable. But I suppose that wasn't intentional...

File: tools/docker/tests/test_flows.sh [hunk 7]
```diff
@@ -91,14 +91,12 @@ test_channel_selection() {
     return $check_error
 }
 
-test_ap_capability_info_reporting() {
-    #TODO: Implement
-    return 1
-}
-
-test_client_capability_info_reporting() {
```
Comment (arnout) [inline tools/docker/tests/test_flows.sh]: Was this function removed intentionally?

**Review by arnout — CHANGES_REQUESTED** [review]
I can fix the issue myself, if you confirm that the rename and the removal of functions were accidental.

File: tools/docker/tests/test_flows.sh (hunk 7 above)
Comment (Galvien) [reply tools/docker/tests/test_flows.sh]: Yes, the weirdly named functions were renamed to `ap_capability_query` and `client_capability_query`.

File: tools/docker/tests/test_flows.sh (hunk 6 above)
Comment (Galvien) [reply tools/docker/tests/test_flows.sh:105]: The ALL variable is fine but I do in fact need to fix `usage()` and I'll do that now.

File: tools/docker/tests/test_flows.sh [hunk 8]
```diff
@@ -102,10 +102,6 @@ test_client_capability_query() {
      #TODO: Implement
     return 1
 }
-test_link_metric_collection() {
```
Comment (Galvien) [inline tools/docker/tests/test_flows.sh:111]: This was a misnamed test, the proper test name is `combined_infra_metrics`.

File: tools/docker/tests/test_flows.sh [hunk 9]
```diff
@@ -159,14 +155,12 @@ usage() {
     echo "      ap_config_renew - AP configuration renew test"
     echo "      ap_config_bss_tear_down - AP configuration BSS Tear Down test"
     echo "      channel_selection - Channel Selection test"
-    echo "      ap_capability_info_reporting - AP Capability info reporting test"
-    echo "      client_capability_info_reporting - Capability info reporting test"
```
Comment (Galvien) [inline tools/docker/tests/test_flows.sh:176]: These two tests also don't belong

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 10]
```diff
@@ -1560,6 +1561,9 @@ bool main_thread::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
         return handle_1905_discovery_query(cmdu_rx);
         //LOG(INFO) << "I got the Topology Query message!";
     }
+    case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE: {
```
Comment (morantr) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I mention it on #161, this handler should not be here, it should be in son_slave.
It doesn't matter that it doesn't have TLVs. Channel preference query is also a message without TLVs and being handled in the son_slave which is responsible for the radio.
The backhaul manager automatically forwards messages to the son_slaves if not handled in the backhaul manager itself.
I don't see the point of handling it here and then create new CMDU to send it to the son_slave.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 11]
```diff
@@ -4696,6 +4698,13 @@ bool slave_thread::parse_non_intel_join_response(Socket *sd)
     return true;
 }
 
+bool slave_thread::handle_1905_ap_capability_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received AP_CAPABILITY_QUERY_MESSAGE" << std::dec << int(mid);
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: Ok AP_CAPABILITY_QUERY_MESSAGE was received, then what?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This is wrong, as said before, the handler should not be here.
CMDUs are passed automatically to the son_slaves.

**Review by morantr — CHANGES_REQUESTED** [review]

**Review by morantr — APPROVED** [review]

**Review by tomereli — APPROVED** [review]
please rebase-squash then we can merge

**Review by arnout — APPROVED** [review]
I made a few small improvements to the commit messages.

**Pull request merged**